File: storage/ndb/test/crund/src/crundndb/Driver.cpp
/* -*- mode: java; c-basic-offset: 4; indent-tabs-mode: nil; -*-
 *  vim:expandtab:shiftwidth=4:tabstop=4:smarttab:
 *
 *  Copyright (c) 2010, 2011, Oracle and/or its affiliates. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; version 2 of the License.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <vector>
#include <cassert>
#include <ctime>

#include "helpers.hpp"
#include "string_helpers.hpp"

#include "Driver.hpp"

using std::cout;
using std::flush;
using std::endl;
using std::ios_base;
using std::ofstream;
using std::ostringstream;
using std::string;
using std::wstring;
using std::vector;

using utils::Properties;
using utils::toBool;
using utils::toInt;
using utils::toString;

//---------------------------------------------------------------------------

vector< string > Driver::propFileNames;
string Driver::logFileName;

void
Driver::exitUsage()
{
    cout << "usage: [options]" << endl
         << "    [-p <file name>]...    properties file name" << endl
         << "    [-l <file name>]       log file name for data output" << endl
         << "    [-h|--help]            print usage message and exit" << endl
         << endl;
    exit(1); // return an error code
}

void
Driver::parseArguments(int argc, const char* argv[])
{
    for (int i = 1; i < argc; i++) {
        const string arg = argv[i];
        if (arg.compare("-p") == 0) {
            if (i >= argc) {
                exitUsage();
            }
            propFileNames.push_back(argv[++i]);
        } else if (arg.compare("-l") == 0) {
            if (i >= argc) {
                exitUsage();
            }
            logFileName = argv[++i];
        } else if (arg.compare("-h") == 0 || arg.compare("--help") == 0) {
            exitUsage();
        } else {
            cout << "unknown option: " << arg << endl;
            exitUsage();
        }
    }

    if (propFileNames.size() == 0) {
        propFileNames.push_back("run.properties");
    }

    if (logFileName.empty()) {
        logFileName = "log_";

        // format, destination strings (static size)
        const char format[] = "%Y%m%d_%H%M%S";
        const int size = sizeof("yyyymmdd_HHMMSS");
        char dest[size];

        // get time, convert to timeinfo (statically allocated) then to string
        const time_t now = time(0);
        const int nchars = strftime(dest, size, format, localtime(&now));
        assert(nchars == size-1);
        (void)nchars;

        logFileName += dest;
        logFileName += ".txt";
    }
    //cout << "logFileName='" << logFileName << "'" << endl;
}

// ----------------------------------------------------------------------

void
Driver::run() {
    init();

    if (nRuns > 0) {
        cout << endl
             << "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++" << endl
             << "hot runs ..." << endl
             << "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++" << endl;

        for (int i = 0; i < nRuns; i++) {
            runTests();
        }

        // write log buffers
        if (logRealTime) {
            log << descr << ", rtime[ms]"
                << header.rdbuf()->str() << endl
                << rtimes.rdbuf()->str() << endl << endl << endl;
        }        
        if (logCpuTime) {
            log << descr << ", ctime[ms]"
                << header.rdbuf()->str() << endl
                << ctimes.rdbuf()->str() << endl << endl << endl;
        }
    }

    close();
}

void
Driver::init() {
    loadProperties();
    initProperties();
    printProperties();
    openLogFile();

    // clear log buffers
    logHeader = true;
    header.rdbuf()->str("");
    rtimes.rdbuf()->str("");
}

void
Driver::close() {
    // clear log buffers
    header.rdbuf()->str("");
    rtimes.rdbuf()->str("");

    closeLogFile();
}

void
Driver::loadProperties() {
    cout << endl;
    for (vector<string>::const_iterator i = propFileNames.begin();
         i != propFileNames.end(); ++i) {
        cout << "reading properties file:        " << *i << endl;
        props.load(i->c_str());
        //cout << "props = {" << endl << props << "}" << endl;
    }
}

void
Driver::initProperties() {
    cout << "setting driver properties ..." << flush;

    ostringstream msg;

    logRealTime = toBool(props[L"logRealTime"], true);
    logCpuTime = toBool(props[L"logCpuTime"], false);

    nRuns = toInt(props[L"nRuns"], 1, -1);
    if (nRuns < 0) {
        msg << "[ignored] nRuns:             '"
            << toString(props[L"nRuns"]) << "'" << endl;
        nRuns = 1;
    }

    //if (msg.tellp() == 0) // netbeans reports amibuities
    if (msg.str().empty()) {
        cout << "   [ok]" << endl;
    } else {
        cout << endl << msg.str() << endl;
    }
}

void
Driver::printProperties() {
    const ios_base::fmtflags f = cout.flags();
    // no effect calling manipulator function, not sure why
    //cout << ios_base::boolalpha;
    cout.flags(ios_base::boolalpha);

    cout << endl << "driver settings ..." << endl;
    cout << "logRealTime:                    " << logRealTime << endl;
    cout << "logCpuTime:                     " << logCpuTime << endl;
    cout << "nRuns:                          " << nRuns << endl;

    cout.flags(f);
}

void
Driver::openLogFile() {
    cout << endl
         << "opening results file:" << flush;
    log.open(logFileName.c_str(), ios_base::out | ios_base::trunc);
    assert(log.good());
    cout << "           [ok: " << logFileName << "]" << endl;
}

void
Driver::closeLogFile() {
    cout << endl
         << "closing results file:" << flush;
    log.close();
    cout << "           [ok: " << logFileName << "]" << endl;
}

// ----------------------------------------------------------------------

void
Driver::begin(const string& name) {
    cout << endl;
    cout << name << endl;

    if (logRealTime && logCpuTime) {
        s0 = hrt_tnow(&t0);
    } else if (logRealTime) {
        s0 = hrt_rtnow(&t0.rtstamp);
    } else if (logCpuTime) {
        s0 = hrt_ctnow(&t0.ctstamp);
    }
}

void
Driver::commit(const string& name) {
    if (logRealTime && logCpuTime) {
        s1 = hrt_tnow(&t1);
    } else if (logRealTime) {
        s1 = hrt_rtnow(&t1.rtstamp);
    } else if (logCpuTime) {
        s1 = hrt_ctnow(&t1.ctstamp);
    }

    if (logRealTime) {
        if (s0 | s1) {
            cout << "ERROR: failed to get the system's real time.";
            rtimes << "\tERROR";
        } else {
            long t = long(hrt_rtmicros(&t1.rtstamp, &t0.rtstamp)/1000);
            cout << "tx real time:                   " << t
                 << "\tms" << endl;
            rtimes << "\t" << t;
            rta += t;
        }
    }

    if (logCpuTime) {
        if (s0 | s1) {
            cout << "ERROR: failed to get this process's cpu time.";
            ctimes << "\tERROR";
        } else {
            long t = long(hrt_ctmicros(&t1.ctstamp, &t0.ctstamp)/1000);
            cout << "tx cpu time:                    " << t
                 << "\tms" << endl;
            ctimes << "\t" << t;
            cta += t;
        }
    }

    if (logHeader)
        header << "\t" << name;
}

//---------------------------------------------------------------------------

File: test/Driver.js
/*
 Copyright (c) 2012, Oracle and/or its affiliates. All rights
 reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; version 2 of
 the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 02110-1301  USA
 */

require("./test_config.js");


/** Driver 
*/
function Driver() {
  this.suites = [];
  this.fileToRun = "";
  this.suitesToRun = "";
};

Driver.prototype.findSuites = function(directory) {
  if(this.fileToRun) {
    var pathname = path.join(directory, this.fileToRun);
    this.suites.push(new harness.Suite("file", pathname));
  }
  else { 
    /* Read the test directory, building list of suites */
    var files = fs.readdirSync(directory);
    for(var i = 0; i < files.length ; i++) {
      var f = files[i];
      var st = fs.statSync(path.join(directory, f));
      if (st.isDirectory() && this.isSuiteToRun(f)) {
        if (debug) console.log('Driver.findSuites found directory ' + f);
        this.suites.push(new harness.Suite(f, path.join(directory, f)));
      }
    }
  }
};

Driver.prototype.isSuiteToRun = function(directoryName) {
  if(directoryName == 'lib') return false;
  if (debug) console.log("SuitesToRun: " + this.suitesToRun);
  if (this.suitesToRun && this.suitesToRun.indexOf(directoryName) == -1) {
    if (debug) console.log('Driver.isSuiteToRun for ' + directoryName + ' returns false.');
    return false;
  }
   if (debug) console.log('Driver.isSuiteToRun for ' + directoryName + ' returns true.');
  return true;
};

Driver.prototype.testCompleted = function(testCase) {
  if (debug) console.log('Driver.testCompleted: ' + testCase.name);
  var suite = testCase.suite;
  if (debug) console.log('Driver.testCompleted suite for ' + suite.name);
  if (suite.testCompleted(testCase)) {
    // this suite is done; remove it from the list of running suites
    if (--this.numberOfRunningSuites == 0) {
      // no more running suites; report and exit
      this.reportResultsAndExit();
    }
  } 
};

Driver.prototype.reportResultsAndExit = function() {
  // close all session factories
  sessionFactories = mynode.getOpenSessionFactories();
  sessionFactories.forEach(function(sessionFactory) {
    if (debug) console.log('Driver.reportResultsAndExit closing ' + sessionFactory.key);
    sessionFactory.close();
  });
  console.log("Passed:  ", result.passed.length);
  console.log("Failed:  ", result.failed.length);
  console.log("Skipped: ", result.skipped.length);
  if(debug) udebug.close();
  process.exit(result.failed.length > 0);  
};


/*****************************************************************************
 ********************** main process *****************************************
 *****************************************************************************/

driver = new Driver();
var result = new harness.Result(driver);
result.listener = new harness.Listener();
var exit = false;
var timeoutMillis = 5000;

var usageMessage = 
  "Usage: node driver [options]\n" +
  "       -h or --help: print this message\n" +
  "      -d or --debug: set the debug flag\n" +
  "      -t or --trace: print stack trace from failing tests\n" +
  "    --suite=<suite>: only run the named suite(s)\n" +
  "   --suites=<suite>: only run the named suite(s)\n" +
  "      --file=<file>: only run the named test file\n" +
  "--adapter=<adapter>: only run on the named adapter (e.g. ndb or mysql)\n" +
  "       --timer=<ms>: set timeout (in msec); set to 0 to disable timer.\n" +
  "--set <var>=<value>: set a global variable, overriding test_config.js\n"
  ;

// handle command line arguments
for(var i = 2; i < process.argv.length ; i++) {
  val = process.argv[i];
  switch (val) {
  case '--debug':
  case '-d':
    console.log('Setting debug to true');
    debug = true;
    udebug.on();
    break;
  case '--help':
  case '-h':
    exit = true;
    break;
  case '--trace':
  case '-t':
    result.listener.printStackTraces = true;
    break;
  case '--set':
    i++;  // next argument
    pair = process.argv[i].split('=');
    if(pair.length == 2) {
      if(debug) console.log("Setting global: " + pair[0] + "=" + pair[1]);
      global[pair[0]] = pair[1];
    }
    else {
      console.log("Invalid --set option " + process.argv[i]);
      exit = true;
    }
    break;
  default:
    values = val.split('=');
    if (values.length == 2) {
      switch (values[0]) {
      case '--suite':
      case '--suites':
        driver.suitesToRun = values[1];
        break;
      case '--file':
        driver.fileToRun = values[1];
        break;
      case '--adapter':
        global.adapter = values[1];
        break;
      case '--timer':
        timeoutMillis = values[1];
        break;
      default:
        console.log('Invalid option ' + val);
        exit = true;
      }
    } else {
      console.log('Invalid option ' + val);
      exit = true;
   }
  }
}

if (exit) {
  console.log(usageMessage);
  process.exit(0);
}


driver.findSuites(__dirname);
if (debug) console.log('suites found ' + driver.suites.length);

driver.suites.forEach(function(suite) {
  if (debug) console.log('createTests for ' + suite.name);
  suite.createTests();
});

// now run tests
driver.numberOfRunningSuites = 0;
driver.suites.forEach(function(suite) {
  if (debug) console.log('main running tests for ' + suite.name);
  if (suite.runTests(result)) {
    driver.numberOfRunningSuites++;
  }
});

// if we did not start any suites, exit now
if (driver.numberOfRunningSuites == 0) {
  driver.reportResultsAndExit();
}

// set a timeout to prevent process from waiting forever
if(timeoutMillis > 0) {
  if (debug) console.log('Setting timeout of ' + timeoutMillis);
  setTimeout(function() {
    var nwait = result.listener.started - result.listener.ended;
    var tests = (nwait == 1 ? " test." : " tests.");
    console.log('TIMEOUT: still waiting for ' + nwait + tests);
  }, timeoutMillis);
}
